tests/torch/quantization/test_functions.py

```diff
@@ -317,48 +474,101 @@ def test_quantize_asymmetric_forward(self, _seed, input_size, bits, use_cuda, is
                 pytest.skip("Skipping CUDA test cases for CPU only setups")
             skip_if_half_on_cpu(is_fp16, use_cuda)
+            if is_fp16:
+                np_dtype = np.float16
+            else:
+                np_dtype = np.float32
             level_low, level_high, levels = self.get_range_level(bits)
-            ref_input = generate_input(input_size)
+            fixed = None
             if is_fp16:
-                ref_input = ref_input.astype(np.float16)
-
-            ref_input_low, ref_input_range = self.generate_range(ref_input, scale_mode, is_weights,
-                                                                 is_fp16)
-            test_input, test_input_low, test_input_range = get_test_data(
-                [ref_input, ref_input_low, ref_input_range], use_cuda, is_backward=True, is_fp16=is_fp16)
+                # This is needed to make scale == 1 to prevent
+                # quant movement on forward pass in FP16 precision.
+                # In case scale != 1., not precice scale multiplication in FP16
+                # could lead to big deviations, so even if an input point
+                # lies in safe range (far from middles of quants) after a scaling
+                # it could end up in the middle of a quant. It happens mostly
+                # when target quant > 150 because in real life scenarious quantization range
+                # usualy less than 2 ** quantization bits,
+                # so input is small and scale is big, small FP16 input multiplies big fp16 scale,
+                # deviation is significant.
+                fixed = {}
+                fixed['input_low'] = - 2 ** (bits - 1)
+                fixed['input_range'] = 2 ** bits - 1
+            ref_input_low, ref_input_range = self.generate_range(input_size, scale_mode, is_weights,
+                                                                 is_fp16, fixed)
+            test_input_low, test_input_range = get_test_data(
+                [ref_input_low, ref_input_range], use_cuda, is_backward=True, is_fp16=is_fp16)
 
             range_sign = np.sign(ref_input_range)
             ref_input_range = abs(ref_input_range) + EPS
             ref_input_low, ref_input_range = ReferenceQuantize.tune_range(
                 ref_input_low, ref_input_range, levels)
+
+            for tensor_ in (test_input_low, test_input_range):
+                assert tensor_.dtype == torch.half if is_fp16 else torch.float
+
+            # This is needed to prevent middle points in forward pass.
+            # Small deviation in computations could lead to completely different
+            # results: instead of quant a quant a + 1 will be returned.
+            # Different results in forward leads to high deviations on bacward pass.
+            # To prevent this, input values are put into [min_deviation, max_deviation]
+            # section of quant, so small deviation woun't change the quant on forward pass
```

Comment:
```suggestion
            # Different results in forward leads to high deviations on backward pass.
            # To prevent this, input values are put into [min_deviation, max_deviation]
            # section of quant, so small deviation won't change the quant on forward pass
```
the same

Reply:
Fixed